src/openforms/utils/json_logic.py

```diff
@@ -49,3 +52,174 @@ def normalize_value(value: Union[dict, list, tuple, int, str]) -> JSONLogicValue
         return JsonLogicTest.from_expression(value)
 
     raise NotImplementedError(f"Unknown value type: {type(value)}")
+
+
+@dataclass
+class ComponentMeta:
+    form_step: "FormStep"
+    component: Dict[str, Any]
+
+
+@dataclass
+class IntrospectionResult:
+
+    expression: dict
+    rule_tree: Dict[Any, str]
+
+    def as_string(self) -> str:
+        return stringify_json_tree(self.rule_tree)
+
+    def get_input_components(self):
+        return get_leaves(self.rule_tree)
+
+
+def introspect_json_logic(
+    expression: Dict[Any, str],
+    components_map: Dict[Any, ComponentMeta],
```

Comment:
```suggestion
    components_map: Dict[str, ComponentMeta],
```
this is a mapping of component key to meta, so should be `str` rather than `Any` (since the keys are strings).